clang/runtime/dpct-rt/include/dpct/math.hpp

```diff
@@ -2216,6 +2218,143 @@ void ldmatrix(uintptr_t addr, T *m1, T *m2, T *m3, T *m4, bool trans = false) {
+  auto a = reinterpret_cast<ABType *>(a_mat);
+  auto b = reinterpret_cast<ABType *>(b_mat);
+  auto c = reinterpret_cast<CDType *>(c_mat);
+
+  auto sg = sycl::ext::oneapi::this_work_item::get_sub_group();
+  int lane = sg.get_local_linear_id();
+
+  static_assert(M == 16 && N == 8 && K == 16,
+                "Only m16n8k16 shape is supported!");
+
+  short ROW_LOAD_OFFSET = 4 * (lane >> 2);
+  short COL_LOAD_OFFSET = 8 * (lane % 4);
+
+  if constexpr (M == 16 && N == 8 && K == 16) {
+    if constexpr (std::is_floating_point_v<CDType>) {
+      // Init D matrix with elements of C matrix
+      *d[0] = c[0];
+      *d[1] = c[1];
+      *d[2] = c[2];
+      *d[3] = c[3];
+
+      // Iterate through 4 neighbouring work items to gather the A & B matix
+      // elements of the rows and cols associated with each work item
+      // WI0: { row0: a0 .. a15 & row8: a0 .. a15 } and
+      //      { col0: b0 .. b15 & col1: b0 .. b15 }
+      for (int i = 0; i < 4; i++) {
+        ABType recv_a[4], recv_b[4];
+
+        // WI0 loads row0: { a0, a1 }, { a2, a3 }, { a4, a5 }, { a6, a7 }
+        recv_a[0] = dpct::select_from_sub_group(sg, a[0], ROW_LOAD_OFFSET + i);
+        // WI0 loads row0: { a8, a9 }, { a10, a11 }, { a12, a13 }, { a14, a15 }
+        recv_a[1] = dpct::select_from_sub_group(sg, a[2], ROW_LOAD_OFFSET + i);
+        // WI0 loads row8: { a0, a1 }, { a2, a3 }, { a4, a5 }, { a6, a7 }
+        recv_a[2] = dpct::select_from_sub_group(sg, a[1], ROW_LOAD_OFFSET + i);
+        // WI0 loads row8: { a8, a9 }, { a10, a11 }, { a12, a13 }, { a14, a15 }
+        recv_a[3] = dpct::select_from_sub_group(sg, a[3], ROW_LOAD_OFFSET + i);
+
+        // WI0 loads col0: { b0, b1 }, { b2, b3 }, { b4, b5 }, { b6, b7 }
+        recv_b[0] = dpct::select_from_sub_group(sg, b[0], COL_LOAD_OFFSET + i);
+        // WI0 loads col0: { b8, b9 }, { b10, b11 }, { b12, b13 }, { b14, b15 }
+        recv_b[1] = dpct::select_from_sub_group(sg, b[1], COL_LOAD_OFFSET + i);
+        // WI0 loads col1: { b0, b1 }, { b2, b3 }, { b4, b5 }, { b6, b7 }
+        recv_b[2] =
+            dpct::select_from_sub_group(sg, b[0], COL_LOAD_OFFSET + 4 + i);
+        // WI0 loads col1: { b8, b9 }, { b10, b11 }, { b12, b13 }, { b14, b15 }
+        recv_b[3] =
+            dpct::select_from_sub_group(sg, b[1], COL_LOAD_OFFSET + 4 + i);
+
+        auto ra = reinterpret_cast<MulType *>(recv_a);
+        auto rb = reinterpret_cast<MulType *>(recv_b);
+
+        // Calculate partial product of 4 A & B matrix elements
+        // For each iteration of i, work-item calculates D matrix values as
+        // below:
+        // d0 += row0{ a0, a1, a8, a9 } * col0{ b0, b1, b8, b9 }
+        // d1 += row0{ a0, a1, a8, a9 } * col1{ b0, b1, b8, b9 }
+        // d2 += row8{ a0, a1, a8, a9 } * col0{ b0, b1, b8, b9 }
+        // d3 += row8{ a1, a1, a8, a9 } * col1{ b0, b1, b8, b9 }
+        for (int j = 0; j < 4; j++) {
+          *d[0] += static_cast<CDType>(ra[j]) * static_cast<CDType>(rb[j]);
```

Comment:
d0~d3 is the four results of result D (`D=A*B+C`),  from the algorithm of matrix multiplication,  d0(e.g., the position of d0 in matrix D is `[i, j]`) is the accumulation of dot multiplication of the whole `i` row of matrix A, and the whole `j` column of matrix B, in the subgroup level, for the current work item, pls explain how the e `i` row of matrix A, and the whole `j` column of matrix B are accessed.

Reply:
Changed the description to reflect Added more description to the algo functionality